### DISCUSSION TITLE
Add Objects (variadic) syntax (similar to classnames, clxs)

### DISCUSSION THREAD

**Discussion by MrEfrem** [dcastil/tailwind-merge#137] [Ideas]
`classnames`, `clxs` libraries support very useful object syntax to conditionally apply rules. 
Would it be possible to implement the same syntax here? Then we will be able to use only one library for class names managing. 

```js
// Objects (variadic)
clsx({ foo:true }, { bar:false }, null, { '--foobar':'hello' });
//=> 'foo --foobar'
```

**Comment by dcastil** [comment]
Hey @MrEfrem! 👋 

I didn't implement the object syntax because I think it's easier to read code using the `&&` syntax instead. Therefore I don't want to add the runtime overhead for object syntax support.

# Long explanation

When I read a collection of conditionally applied classes, I'm always interested in the condition first to know whether I even need to think about the classes or not and only afterwards I want to deal with the actual classes. The object notation turns that around and increases the cognitive load because I need to skip the string with all the classes to get to the condition and then return back to the string again.

Let me show you what I mean on a real example (formatted with [Prettier](https://prettier.io)), first using an object.

```js
clsx({
    'px-8 sticky grid items-center grid-flow-col grid-cols-1 transition cursor-pointer bg-clip-padding border-y border-white hover:bg-grey-700-opaque-5':
        true,
    'bg-white': !transparentWhenFloaty || position !== Position.Floaty,
    invisible: isInvisible,
    // Fixes sticky position bug in Chrome. More info: <private link>
    'before:absolute before:left-0 before:-top-px before:w-full before:h-px before:bg-white':
        position === Position.StickyTop,
    'before:absolute before:left-0 before:-bottom-px before:w-full before:h-px before:bg-white':
        position === Position.StickyBottom,
    'border-t-grey-700/10': position === Position.StickyBottom,
    'border-b-grey-700/10': position !== Position.StickyBottom,
})
```

Let's say I need to fix something at this DOM node when the component is in the StickyBottom position and want to check all the classes being applied in this state. Here are some aspects I don't like about this syntax:
1. While reading the code, I constantly need to jump back and forth, skipping the classes string, checking the condition and going back to the classes string to acknowledge the classes as being applied.
2. Because the class strings have very different lengths, every condition is in a different place which makes it harder to find quickly.
3. It's easy to miss the `invisible` class because it doesn't have quotes around it and therefore editors apply different styling to it. E.g. I'm used to my editor displaying strings in blue and focus on that color, but the `invisible` class is rendered in white. ![Screenshot of code snippet above in my editor](https://user-images.githubusercontent.com/31006608/186864151-9c966a79-6231-4410-a57e-ed22b218c2c2.png)

Now let's take a look at the same code using the logical `&&` operator instead. This is also the original code.

```js
clsx(
    'px-8 sticky grid items-center grid-flow-col grid-cols-1 transition cursor-pointer bg-clip-padding border-y border-white hover:bg-grey-700-opaque-5',
    (!transparentWhenFloaty || position !== Position.Floaty) && 'bg-white',
    isInvisible && 'invisible',
    // Fixes sticky position bug in Chrome. More info: <private link>
    position === Position.StickyTop &&
        'before:absolute before:left-0 before:-top-px before:w-full before:h-px before:bg-white',
    position === Position.StickyBottom &&
        'before:absolute before:left-0 before:-bottom-px before:w-full before:h-px before:bg-white',
    position === Position.StickyBottom ? 'border-t-grey-700/10' : 'border-b-grey-700/10'
)
```

Thinking about the three points mentioned:
1. I don't need to jump back and forth in code anymore when reading it. I can read the condition and when it isn't interesting to me, I skip the following string and continue to the next condition.
2. Every new line starts with a condition which makes it easier for me to find the ones I'm interested in.
3. All class strings, including the one containing `invisible` have the same styling in my editor which makes it faster to find them.

# Conclusion

I think the object syntax shouldn't be used to apply Tailwind classes conditionally and therefore I didn't add support for it in tailwind-merge. Because of this I built a [~~`join`~~](https://github.com/dcastil/tailwind-merge/blob/v1.6.0/docs/api-reference.md#join) [`twJoin`](https://github.com/dcastil/tailwind-merge/blob/v1.9.1/docs/api-reference.md#twjoin) function which is essentially [a copy of clsx](https://github.com/dcastil/tailwind-merge/blob/v1.6.0/src/lib/join.ts) but without support for objects which makes it even faster than clsx.

But if you have a use case where you think the object syntax makes more sense than `&&`, I'm interested in hearing about it! There's always the chance that I missed something.

---

Related: https://github.com/dcastil/tailwind-merge/discussions/121#discussioncomment-3381696, https://github.com/dcastil/tailwind-merge/issues/124

**Reply by MrEfrem** [reply]
Thanls @dcastil  for this great explanation.

But on my opinion this approach looks better

```js
clsx(
    'px-8 sticky grid items-center grid-flow-col grid-cols-1 transition'
    'cursor-pointer bg-clip-padding border-y border-white',
    'hover:bg-grey-700-opaque-5 py-4 mb-10 rounded-lg',
    {
        'bg-white': !transparentWhenFloaty || position !== Position.Floaty,
        invisible: isInvisible,
        // Fixes sticky position bug in Chrome. More info: <private link>
        'before:absolute before:left-0 before:-top-px before:w-full before:h-px before:bg-white': position === Position.StickyTop,
        'before:absolute before:left-0 before:-bottom-px before:w-full before:h-px before:bg-white': position === Position.StickyBottom,
        'border-t-grey-700/10': position === Position.StickyBottom
        'border-b-grey-700/10': position !== Position.StickyBottom
    }
)
```

than this

```js
clsx(
    'px-8 sticky grid items-center grid-flow-col grid-cols-1 transition'
    'cursor-pointer bg-clip-padding border-y border-white',
    'hover:bg-grey-700-opaque-5 py-4 mb-10 rounded-lg',
    (!transparentWhenFloaty || position !== Position.Floaty) && 'bg-white',
    isInvisible && 'invisible',
    // Fixes sticky position bug in Chrome. More info: <private link>
    position === Position.StickyTop &&
        'before:absolute before:left-0 before:-top-px before:w-full before:h-px before:bg-white',
    position === Position.StickyBottom &&
        'before:absolute before:left-0 before:-bottom-px before:w-full before:h-px before:bg-white',
    position === Position.StickyBottom ? 'border-t-grey-700/10' : 'border-b-grey-700/10'
)
```

because you see your classes first and you usually look at Figma, your browser or something more where you can take your design. So for you it's no
less important to know which TW classes you need to add, change, delete.
But when you see mix classes with conditions in the second approach it's really more difficult to talk about design.
Finally, I think should be implemented both approaches because for every of them there are pros and cons.

**Reply by dcastil** [reply]
Thanks for your point of view on this! I'll think about this. 🤔 

You can also make twMerge accept objects by wrapping it with your own function for the time being.

```js
import { twMerge as twMergeOriginal } from 'tailwind-merge'
import clsx from 'clsx'

export function twMerge(...args) {
    return twMergeOriginal(clsx(args))
}
```

**Reply by marcelo-mason** [reply]
Your open-minded reply was refreshing. Thanks for providing a workaround. I prefer a mixture where I use object syntax to group related conditions together.

**Reply by SergeITinc** [reply]
Hey @dcastil! 👋

Thank you for your detailed explanation regarding the object syntax and its impact on readability and cognitive load. I understand your perspective and appreciate the thought process behind preferring the && syntax.

@MrEfrem 
However, I believe there's a middle ground that maintains readability without introducing the runtime overhead of supporting object syntax. 
Here's my proposal using array notation with twMerge:
```tsx
        className={twMerge(
          "sticky grid grid-flow-col grid-cols-1 items-center px-8 transition",
          "cursor-pointer border-y border-white bg-clip-padding",
          "hover:bg-grey-700-opaque-5 mb-10 rounded-lg py-4",
          [
            (!transparentWhenFloaty || position !== Position.Floaty) &&
              "bg-white",
          ],
          [isInvisible && "invisible"],
          // Fixes sticky position bug in Chrome. More info: <private link>
          [
            position === Position.StickyTop &&
              "before:absolute before:-top-px before:left-0 before:h-px before:w-full before:bg-white",
            position === Position.StickyBottom &&
              "before:absolute before:-bottom-px before:left-0 before:h-px before:w-full before:bg-white",
            position === Position.StickyBottom
              ? "border-t-grey-700/10"
              : "border-b-grey-700/10",
          ],
        )}
```
This approach keeps the classes visible upfront, making it easier to see which Tailwind classes are being applied, while still conditionally applying classes based on the state. It avoids the cognitive load of jumping back and forth between conditions and class names.

I think this strikes a good balance between readability and performance, and it doesn't require adding support for object syntax.

Looking forward to hearing your thoughts!

Best regards,
Serge

**Reply by dcastil** [reply]
Hey @SergeITinc! 👋

The array syntax already works in tailwind-merge! It's supported since [v1.6.0](https://github.com/dcastil/tailwind-merge/releases/tag/v1.6.0).

**Reply by SergeITinc** [reply]
@dcastil

Thank you for the quick response!

I apologize if my previous message was unclear. I wasn't suggesting creating a new approach or introducing object syntax. My intention was to demonstrate how we can effectively use the existing array notation in tailwind-merge to handle conditional class names.


This approach leverages the array syntax to conditionally apply classes without adding any new features or syntaxes. It keeps the code clean and readable while utilizing the capabilities of tailwind-merge.

My goal was to show different ways to use the existing functionality for anyone reading the discussion.

Thank you for your understanding!

Best regards,
@SergeITinc

**Reply by dcastil** [reply]
Ah got it. 🙂

**Reply by alexchexes** [reply]
Hello @dcastil! I just wanted to share a thought about what you said:

>3. It's easy to miss the `invisible` class because it doesn't have quotes around it and therefore editors apply different styling to it. E.g. I'm used to my editor displaying strings in blue and focus on that color, but the invisible class is rendered in white.

I completely agree, and it’s definitely frustrating. However, I always use `"quoteProps": "preserve"` in my `.prettierrc` to avoid the unwanted removal of quotes when they’re actually needed. Is there a particular reason why you don’t do the same? Just curious. Thank you for your work!

```jsonc
// .prettierrc.json
{
  "$schema": "https://json.schemastore.org/prettierrc",
  "semi": false,
  "singleQuote": true,
  "printWidth": 100,
  "quoteProps": "preserve", // here it is
}
```

**Reply by dcastil** [reply]
@alexchexes oh that's a nice setting, didn't know about it. Might apply this, too. But my point about jumps in reading code still stands.

**Reply by alexchexes** [reply]
Yeah, I understand, no arguing. It's your project and your decision. Anyway, for those of us who are used to object syntax and comfortable with it, we can indeed wrap twMerge in a helper like `const twMergeX = (...args) => twMerge(clsx(args))`

**Reply by GiantRobots** [reply]
Hi there! I understand the style preference but it unfortunately conflicts with Svelte5's usage of cslx. It would be really convenient to just be able to do....

```svelte
<script lang="ts">
	import { twMerge } from 'tailwind-merge';
	import type { ClassValue } from 'svelte/elements';
	import type { Snippet } from 'svelte';

	let { class: className, children }: { class: ClassValue; children: Snippet } =
		$props();
</script>

<button class={twMerge(className)}>
	{@render children?.()}
</button>
```

But we get the type error of

```
Type 'ClassValue' is not assignable to type 'ClassNameValue'.
  Type 'ClassArray' is not assignable to type 'ClassNameValue'.
    Type 'ClassValue[]' is not assignable to type 'ClassNameArray'.
      Type 'ClassValue' is not assignable to type 'ClassNameValue'.
        Type 'number' is not assignable to type 'ClassNameValue'.
```
        
To fix this I will likely do what @alexchexes recommended which is add cslx to our twMerge for the functionality but it's unfortunate to have to have our own implementation when it feels like it should be a slam dunk to just use twMerge directly on classNames.

**Reply by dcastil** [reply]
@GiantRobots you can create a file with your own twMerge which allows these types.

Either

```ts
import { twMerge as twMergeOriginal } from 'tailwind-merge'

// Just wing it and promise not to use objects as inputs
export const twMerge = twMergeOriginal as (...args: any[]) => string
```

or

```ts
import { twMerge as twMergeOriginal } from 'tailwind-merge'
import { clsx } from 'clsx'

export function twMerge(...args: any[]) {
    return twMergeOriginal(clsx(args))
}
```

**Reply by GiantRobots** [reply]
Yeah I'll be doing that in our code but wanted to let you/others know about that piece of interop weirdness as a result of this decision.

**Reply by armchair-traveller** [reply]
Svelte again. I also avoid object syntax, but honestly these types feel more trouble than they’re worth. I don’t want to install `clsx` just for its types when Svelte already pulls it in--that’s duplicate deps for no real benefit. And with `tailwind-merge` having its own flavor, that’s effectively three copies of the same thing, haha.

The “easy” fixes don’t work with **Tailwind Variants** because it depends on `ClassNameValue` from `tailwind-merge` as a peer dep. For example:

```ts
buttonVariants({ className })
```

`className` throws similar errors to [[@GiantRobots’ issue](https://github.com/dcastil/tailwind-merge/discussions/137#discussioncomment-13534704)](https://github.com/dcastil/tailwind-merge/discussions/137#discussioncomment-13534704).

Unlike `twMerge`, which you can patch with a simple `declare module`/`declare const`, `tv`’s types are a web of deep generics and not straightforward to override. I’m sure a TypeScript wizard could prove me wrong, haha...but I’m not eager to wrap `tv` and ship my own module just to make TS happy.

For now, I’m using the `ClassProp` type that `tailwind-variants` exposes and intersecting it onto my props whenever `tv` is used, not often does a component have need for variants after all.

Not begging for a fix, just adding another perspective.
